### PR TITLE
[DRAFT, NOT FOR MERGE] fix sq_quote_buf_pretty

### PR DESCRIPTION
**Pull request by garimasi514**
fix, tests: make sure sq_quote_buf_pretty doesn't inadvertently drop tokens that are empty and null, add tests for sq_quote_buf_pretty (quote.c) around some basic scenarios and to document the behavior of the method for null and empty strings

The only test failure is "not ok 48 - detect bogus diffFilter output" This has probably been fixed in git-for-windows but not microsoft/git. Completely unrelated to the changes here.

### CODE REVIEW

**Comment by derrickstolee** [comment]
@garimasi514 It looks like you are using master from git-for-windows/git! You'll want to rebase onto the merge base in common with every version, say v2.22.0.

File: t/t0091-sq-quote-buf-pretty.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#!/bin/sh
```
Comment (derrickstolee) [inline t/t0091-sq-quote-buf-pretty.sh]:
```
git chmod a+x t/t0091-sq-quote-buf-pretty.sh
git commit -a --amend
```

Should fix your build break.

Comment (derrickstolee) [reply t/t0091-sq-quote-buf-pretty.sh]: IGNORE ABOVE I'M AN IDIOT:

```
git add --chmod=a+x t/t0091-sq-quote-buf-pretty.sh
git commit --amend
```

https://git-scm.com/docs/git-add#Documentation/git-add.txt---chmod-x

File: quote.c [hunk 2]
```diff
@@ -47,7 +47,16 @@ void sq_quote_buf_pretty(struct strbuf *dst, const char *src)
 {
 	static const char ok_punct[] = "+,-./:=@_^";
 	const char *p;
-
+	
+	/* 
+	In case of null or empty strings, add a '' to ensure we 
+	don't inadvertently drop these tokens
+	*/
```
Comment (jeffhostetler) [inline quote.c]: super-nit:  they like comments like this to look like

    /*
     * In case ...
     * don't ...
     */

File: t/helper/test-quote.c [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#include "strbuf.h"
+#include "string.h"
+
+int cmd__quote_buf_pretty(int argc, const char **argv)
+{
+	struct strbuf buf_payload = STRBUF_INIT;
+	char* nullString = 0;
+	char empty[5] = {'\0'};
+	
+	if (!strcmp(argv[1], "nullString")){
```
Comment (jeffhostetler) [inline t/helper/test-quote.c]: super-nit:  a single space between rparen and the lcurly.

File: t/helper/test-quote.c [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#include "test-tool.h"
+#include "quote.h"
+#include "strbuf.h"
+#include "string.h"
+
+int cmd__quote_buf_pretty(int argc, const char **argv)
+{
+	struct strbuf buf_payload = STRBUF_INIT;
+	char* nullString = 0;
+	char empty[5] = {'\0'};
```
Comment (jeffhostetler) [inline t/helper/test-quote.c]: is "empty" used anywhere?  i didn't see it.

File: t/helper/test-quote.c [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
```
Comment (jeffhostetler) [inline t/helper/test-quote.c]: so argv[0] is the name of the test "quote_buf_pretty".  and argv[1] is the string to test.

we'll crash if they omit a test string.  that is, if they just type "test-tool quote_buf_pretty".

File: t/helper/test-quote.c [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+int cmd__quote_buf_pretty(int argc, const char **argv)
+{
+	struct strbuf buf_payload = STRBUF_INIT;
+	char* nullString = 0;
+	char empty[5] = {'\0'};
+	
+	if (!strcmp(argv[1], "nullString")){
+		strbuf_addstr(&buf_payload, ".");
+		sq_quote_buf_pretty(&buf_payload, nullString);
+		printf("%s.\n", buf_payload.buf);
```
Comment (jeffhostetler) [inline t/helper/test-quote.c]: This construction is a little confusing and might meet with a little resistance upstream.
I think my concern is that you're using the 2 dots as boundary markers to help make
the shell test script easier.  And that's fine.  I think moving the dots to the printf makes
it clearer that they are just that and not part of the buffer.
Maybe something like:

    sq_quote_buf_pretty(&buf_payload, whatever);
    printf(".%s.\n", buf_payload.buf);

Comment (jeffhostetler) [reply t/helper/test-quote.c]: maybe also have a short comment to that effect.

File: t/helper/test-quote.c [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+		return 0;
+	}
+	
+	if (!strcmp(argv[1], "")){
+		strbuf_addch(&buf_payload, '.');
+		sq_quote_buf_pretty(&buf_payload, empty);
+		printf("%s.\n", buf_payload.buf);
```
Comment (jeffhostetler) [inline t/helper/test-quote.c]: same here

File: t/helper/test-tool.c [hunk 8]
```diff
@@ -57,6 +57,7 @@ static struct test_cmd cmds[] = {
 	{ "sha1-array", cmd__sha1_array },
 	{ "sha256", cmd__sha256 },
 	{ "sigchain", cmd__sigchain },
+	{ "quote_buf_pretty", cmd__quote_buf_pretty },
```
Comment (jeffhostetler) [inline t/helper/test-tool.c]: for some obscure reason this project likes dashes rather than underscores.
(conservation of shift-keys maybe -- i don't know)
anyway, maybe let the command name be "quote-buf-pretty".

File: t/helper/test-quote.c [hunk 9]
```diff
@@ -0,0 +1,30 @@
+#include "test-tool.h"
+#include "quote.h"
+#include "strbuf.h"
+#include "string.h"
+
+int cmd__quote_buf_pretty(int argc, const char **argv)
+{
+	struct strbuf buf_payload = STRBUF_INIT;
+	char* nullString = 0;
```
Comment (jeffhostetler) [inline t/helper/test-quote.c]: you can spell this as

    char *nullString = NULL;

also, the '*' should usually touch the variable not the base type.  (nit nit nit :-)

File: t/helper/test-quote.c (hunk 5 above)
Comment (jeffhostetler) [reply t/helper/test-quote.c]: so maybe have something like:

    if (!argv[1])
        die("missing value....");

File: t/t0091-sq-quote-buf-pretty.sh [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	cat >expect <<-EOF &&
+	'\'test~String\''
+	EOF
+	test-tool quote_buf_pretty test~String >actual &&
+	test_cmp expect actual
+'
+
+test_expect_success 'string with punctuation that requires special quotes' '
+	cat >expect <<-EOF &&
+	'\'test\'\\!\'String\''
```
Comment (jeffhostetler) [inline t/t0091-sq-quote-buf-pretty.sh]: I'm curious why this produces inner quotes (between the words and the bang).

File: t/helper/test-quote.c [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+	char empty[5] = {'\0'};
+	
+	if (!argv[1]) {
+		die("missing input string");
+	}
+	
+	if (!strcmp(argv[1], "nullString")) {
+		sq_quote_buf_pretty(&buf_payload, nullString);
+
+		// Cushion the result with '.'s
```
Comment (jeffhostetler) [inline t/helper/test-quote.c]: they hate // comments.  they only want the /* */ forms.

File: t/helper/test-quote.c [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+		
+		// Cushion the result with '.'s 
+		// to make the test script more readable
+		printf(".%s.\n", buf_payload.buf);
+		return 0;
+	}
+	
+	sq_quote_buf_pretty(&buf_payload, argv[1]);
+	printf("%s\n", buf_payload.buf);
+
```
Comment (jeffhostetler) [inline t/helper/test-quote.c]: just for completeness, there should be a strbuf_release(&buf_payload) before all of the returns to avoid leaking memory.  it doesn't matter so much for tests.

File: quote.c [hunk 13]
```diff
@@ -47,7 +47,16 @@ void sq_quote_buf_pretty(struct strbuf *dst, const char *src)
 {
 	static const char ok_punct[] = "+,-./:=@_^";
 	const char *p;
-
+
```
Comment (derrickstolee) [inline quote.c]: this newline makes me think that you've been writing with CRLFs. `git log --check` will help find out.

File: t/t0091-sq-quote-buf-pretty.sh [hunk 14]
```diff
@@ -0,0 +1,58 @@
+#!/bin/sh
+
+test_description='Testing the sq_quote_buf_pretty method in quote.c'
```
Comment (derrickstolee) [inline t/t0091-sq-quote-buf-pretty.sh]: This makes me think that the test file should be named for the "quote" file. Say, `t0091-quote.sh`. Yes, you only include tests for that method, but if we want to come in and carefully test quoting logic, this is where it would go.

**Comment by derrickstolee** [comment]
@garimasi514 this one looks like it is still failing due to the non-executable test script:

```
non-executable tests: t0091-quote.sh
```

Did we figure out how to do the chmod in Git?

**Review by derrickstolee — APPROVED** [review]
This is ready for GitGitGadget, barring the builds being broken.

**Review by garimasi514 — COMMENTED** [review]
> @garimasi514 this one looks like it is still failing due to the non-executable test script:
> 
> ```
> non-executable tests: t0091-quote.sh
> ```
> 
> Did we figure out how to do the chmod in Git?

We did. I think one of my rebases undid that change. Let me try again. Thanks!

**Comment by garimasi514** [comment]
Will go to the list via gitgitgadget.
